### DISCUSSION TITLE
Why is unit an alias of int in sail.h

### DISCUSSION THREAD

**Discussion by martinberger** [rems-project/sail#356] [General]
What is the rationale behind making `unit` an alias of `int`? In [sail.h](https://github.com/rems-project/sail/blob/sail2/lib/sail.h#L125) we find:
```
 typedef int unit;
```
I would have expected `unit` to be something like `struct unit {};` which, I think, is used in C++.

**Comment by Alasdair** [comment]
`struct unit {}` is valid C++ but not valid C, as empty structs are (as far as I understand) not allowed in C, although some compilers support it as an unofficial extension.

**Reply by martinberger** [reply]
OK. Presumably `typedef void unit;` is also not valid C (but accepted by compilers)?

**Reply by Alasdair** [reply]
If the goal is to have a guaranteed zero-width type, I don't know of a good way to do that in standard C. I think void is a bit too special to be used here, as it's more like a thing with no size rather than a thing with zero size, and it can't really be used as a regular type. Something like `enum unit {UNIT}` might be the closest logically, but I think that would still end up being represented as an int with `UNIT = 0`.

**Comment by PeterSewell** [comment]
On Fri, 27 Oct 2023 at 20:50, Alasdair Armstrong ***@***.***>
wrote:

> If the goal is to have a guaranteed zero-width type, I don't know of a
> good way to do that in standard C. I think void is a bit too special to be
> used here, as it's more like a thing with no size rather than a thing with
> zero size, and it can't really be used as a regular type. Something like enum
> unit {UNIT} might be the closest logically, but I think that would still
> end up being represented as an int with UNIT = 0.
>
Empty structs are not ISO C, but they are allowed in a GCC (and presumably
also Clang) extension:
https://gcc.gnu.org/onlinedocs/gcc-4.1.2/gcc/Empty-Structures.html

I don't know whether there are social or technical reasons why we'd want to
avoid using them for Sail->C, if there's a strong argument for them.

Peter
